--- a/netlink/nss_nl.h
+++ b/netlink/nss_nl.h
@@ -25,6 +25,7 @@
 #define NSS_NL_DEBUG_LVL_WARN 2
 #define NSS_NL_DEBUG_LVL_INFO 3
 #define NSS_NL_DEBUG_LVL_TRACE 4
+#define GENL_ID_GENERATE 0


 #if defined(CONFIG_DYNAMIC_DEBUG)
--- a/netlink/nss_nldtls.c
+++ b/netlink/nss_nldtls.c
@@ -1,12 +1,9 @@
 /*
  **************************************************************************
- * Copyright (c) 2015-2016,2018-2021 The Linux Foundation. All rights reserved.
- * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
- *
+ * Copyright (c) 2015-2016,2018-2020 The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
- *
  * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
@@ -449,7 +446,7 @@ static void nss_nldtls_data_callback(voi
  * nss_nldtls_create_session()
  *	Create a DTLS session through dtlsmgr driver API.
  */
-static struct net_device *nss_nldtls_create_session(struct nss_nldtls_rule *nl_rule)
+static struct net_device *nss_nldtls_create_session(struct nss_nldtls_rule *nl_rule, uint32_t flags)
 {
 	struct nss_nldtls_tun_ctx *dtls_tun_data;
 	struct nss_dtlsmgr_config dcfg;
@@ -466,7 +463,7 @@ static struct net_device *nss_nldtls_cre

 	memset(&dcfg, 0, sizeof(struct nss_dtlsmgr_config));
 	algo = nl_rule->msg.create.encap.cfg.crypto.algo;
-	dcfg.flags = nl_rule->msg.create.flags | NSS_DTLSMGR_ENCAP_METADATA;
+	dcfg.flags = flags | (NSS_DTLSMGR_ENCAP_METADATA | NSS_DTLSMGR_HDR_CAPWAP);
 	if (algo == NSS_DTLSMGR_ALGO_AES_GCM)
 		dcfg.flags |= NSS_DTLSMGR_CIPHER_MODE_GCM;

@@ -608,11 +605,7 @@ static int nss_nldtls_create_ipv4_rule_e
 	ipv4.dest_port = nl_rule->msg.create.encap.cfg.sport;
 	ipv4.dest_port_xlate = nl_rule->msg.create.encap.cfg.sport;

-	if (nl_rule->msg.create.flags & NSS_DTLSMGR_HDR_UDPLITE)
-		ipv4.protocol = IPPROTO_UDPLITE;
-	else
-		ipv4.protocol = IPPROTO_UDP;
-
+	ipv4.protocol = IPPROTO_UDP;
 	ipv4.in_vlan_tag[0] = NSS_NLDTLS_VLAN_INVALID;
 	ipv4.out_vlan_tag[0] = NSS_NLDTLS_VLAN_INVALID;
 	ipv4.in_vlan_tag[1] = NSS_NLDTLS_VLAN_INVALID;
@@ -620,8 +613,6 @@ static int nss_nldtls_create_ipv4_rule_e

 	memcpy(&ipv4.src_mac[0], &nl_rule->msg.create.gmac_ifmac[0], sizeof(ipv4.src_mac));

-	dev_put(ndev);
-
 	/*
 	 * Create an ipv4 rule entry
 	 */
@@ -663,11 +654,7 @@ static int nss_nldtls_create_ipv6_rule_e
 	 */
 	memcpy(ipv6.src_ip, nl_rule->msg.create.encap.cfg.dip, sizeof(ipv6.src_ip));
 	memcpy(ipv6.dest_ip, nl_rule->msg.create.encap.cfg.sip, sizeof(ipv6.dest_ip));
-
-	if (nl_rule->msg.create.flags & NSS_DTLSMGR_HDR_UDPLITE)
-		ipv6.protocol = IPPROTO_UDPLITE;
-	else
-		ipv6.protocol = IPPROTO_UDP;
+	ipv6.protocol = IPPROTO_UDP;

 	ipv6.in_vlan_tag[0] = NSS_NLDTLS_VLAN_INVALID;
 	ipv6.in_vlan_tag[1] = NSS_NLDTLS_VLAN_INVALID;
@@ -676,8 +663,6 @@ static int nss_nldtls_create_ipv6_rule_e

 	memcpy(&ipv6.src_mac[0], &nl_rule->msg.create.gmac_ifmac[0], sizeof(ipv6.src_mac));

-	dev_put(ndev);
-
 	/*
 	 * Create an ipv6 rule entry
 	 */
@@ -744,7 +729,7 @@ static int nss_nldtls_ops_create_tun(str
 	 * Create tunnel based on ip version
 	 */
 	if (nl_rule->msg.create.ip_version == NSS_NLDTLS_IP_VERS_4) {
-		dtls_dev = nss_nldtls_create_session(nl_rule);
+		dtls_dev = nss_nldtls_create_session(nl_rule, NSS_NLDTLS_IPV4_SESSION);
 		if (!dtls_dev) {
 			nss_nl_error("%px: Unable to create dtls session for v4\n", skb);
 			return -EINVAL;
@@ -763,7 +748,7 @@ static int nss_nldtls_ops_create_tun(str
 		atomic_inc(&gbl_ctx.num_tun);
 		nss_nl_info("%px: Successfully created ipv4 dtls tunnel\n", skb);
 	} else {
-		dtls_dev = nss_nldtls_create_session(nl_rule);
+		dtls_dev = nss_nldtls_create_session(nl_rule, NSS_DTLSMGR_HDR_IPV6);
 		if (!dtls_dev) {
 			nss_nl_error("%px: Unable to create dtls session for v6\n", skb);
 			return -EINVAL;
@@ -886,7 +871,6 @@ static int nss_nldtls_ops_update_config(
 	key_len = nl_rule->msg.update_config.config_update.crypto.cipher_key.len;
 	if (key_len > NSS_NLDTLS_CIPHER_KEY_MAX) {
 		nss_nl_error("Invalid cipher length: %u\n", key_len);
-		dev_put(dev);
 		return -EINVAL;
 	}

@@ -894,7 +878,6 @@ static int nss_nldtls_ops_update_config(
 	key_len = nl_rule->msg.update_config.config_update.crypto.auth_key.len;
 	if (key_len > NSS_NLDTLS_AUTH_KEY_MAX) {
 		nss_nl_error("Invalid authentication length: %u\n", key_len);
-		dev_put(dev);
 		return -EINVAL;
 	}

@@ -902,7 +885,6 @@ static int nss_nldtls_ops_update_config(
 	key_len = nl_rule->msg.update_config.config_update.crypto.nonce.len;
 	if (key_len > NSS_NLDTLS_NONCE_SIZE_MAX) {
 		nss_nl_error("Invalid nonce length: %u\n", key_len);
-		dev_put(dev);
 		return -EINVAL;
 	}


--- a/netlink/nss_nlgre_redir_cmn.c
+++ b/netlink/nss_nlgre_redir_cmn.c
@@ -1,6 +1,6 @@
 /*
  ***************************************************************************
- * Copyright (c) 2015-2016, 2018-2021, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2015-2016,2018-2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -325,7 +325,7 @@ static struct rtnl_link_stats64 *nss_nlg
 	int i;

 	for (i = 0; i < NSS_GRE_REDIR_MAX_INTERFACES; i++) {
-		if (!nss_gre_redir_stats_get(i, &get_stats)) {
+		if (!nss_gre_redir_get_stats(i, &get_stats)) {
 			continue;
 		}

@@ -340,15 +340,15 @@ static struct rtnl_link_stats64 *nss_nlg
 	if (found == false)
 		return NULL;

-	stats->tx_bytes   = get_stats.tstats.tx_bytes;
-	stats->tx_packets = get_stats.tstats.tx_packets;
-	stats->rx_bytes   = get_stats.tstats.rx_bytes;
-	stats->rx_packets = get_stats.tstats.rx_packets;
-	for (i = 0;i < ARRAY_SIZE(get_stats.tstats.rx_dropped); i++) {
-		stats->rx_dropped += get_stats.tstats.rx_dropped[i];
+	stats->tx_bytes   = get_stats.node_stats.tx_bytes;
+	stats->tx_packets = get_stats.node_stats.tx_packets;
+	stats->rx_bytes   = get_stats.node_stats.rx_bytes;
+	stats->rx_packets = get_stats.node_stats.rx_packets;
+	for (i = 0;i < ARRAY_SIZE(get_stats.node_stats.rx_dropped); i++) {
+		stats->rx_dropped += get_stats.node_stats.rx_dropped[i];
 	}

-	stats->tx_dropped = get_stats.tstats.tx_dropped;
+	stats->tx_dropped = get_stats.tx_dropped;

 	return stats;
 }
 @@ -384,7 +384,11 @@ static int nss_nlgre_redir_cmn_set_mac_a
 		return -EINVAL;
 	}

+#if (LINUX_VERSION_CODE > KERNEL_VERSION(6, 1, 0))
+	memcpy((void *) dev->dev_addr, addr->sa_data, ETH_ALEN);
+#else
 	memcpy(dev->dev_addr, addr->sa_data, ETH_ALEN);
+#endif
 	return 0;
 }

--- a/netlink/nss_nlipsec.c
+++ b/netlink/nss_nlipsec.c
@@ -1,12 +1,9 @@
 /*
  **************************************************************************
- * Copyright (c) 2015-2016,2018-2021 The Linux Foundation. All rights reserved.
- * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
- *
+ * Copyright (c) 2015-2016,2018-2020 The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
- *
  * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
@@ -52,8 +49,8 @@ static int nss_nlipsec_op_create_tunnel(
 static int nss_nlipsec_op_destroy_tunnel(struct sk_buff *skb, struct genl_info *info);
 static int nss_nlipsec_op_add_sa(struct sk_buff *skb, struct genl_info *info);
 static int nss_nlipsec_op_delete_sa(struct sk_buff *skb, struct genl_info *info);
-static int nss_nlipsec_op_add_flow(struct sk_buff *skb, struct genl_info *info);
-static int nss_nlipsec_op_delete_flow(struct sk_buff *skb, struct genl_info *info);
+// static int nss_nlipsec_op_add_flow(struct sk_buff *skb, struct genl_info *info);
+// static int nss_nlipsec_op_delete_flow(struct sk_buff *skb, struct genl_info *info);

 /*
  * Hold netdevice references
@@ -111,14 +108,14 @@ static struct genl_ops nss_nlipsec_ops[]
 		.cmd = NSS_NLIPSEC_CMD_DEL_SA,
 		.doit = nss_nlipsec_op_delete_sa,
 	},
-	{ /* Add flow */
-		.cmd = NSS_NLIPSEC_CMD_ADD_FLOW,
-		.doit = nss_nlipsec_op_add_flow,
-	},
-	{ /* Delete flow */
-		.cmd = NSS_NLIPSEC_CMD_DEL_FLOW,
-		.doit = nss_nlipsec_op_delete_flow,
-	},
+	// { /* Add flow */
+	// 	.cmd = NSS_NLIPSEC_CMD_ADD_FLOW,
+	// 	.doit = nss_nlipsec_op_add_flow,
+	// },
+	// { /* Delete flow */
+	// 	.cmd = NSS_NLIPSEC_CMD_DEL_FLOW,
+	// 	.doit = nss_nlipsec_op_delete_flow,
+	// },
 };

 /*
@@ -329,7 +326,7 @@ int nss_nlipsec_get_mtu(struct net_devic
 static int nss_nlipsec_op_create_tunnel(struct sk_buff *skb, struct genl_info *info)
 {
 	struct nss_nlipsec_rule *nl_rule;
-	struct nss_ipsecmgr_callback cb = {0};
+	struct nss_ipsecmgr_callback cb;
 	struct nss_nlcmn *nl_cm;
 	struct net_device *dev;
 	struct sk_buff *resp;
@@ -511,8 +508,6 @@ static struct nss_nlipsec_rule *nss_nlip
 		dev_put(*dev);
 		return NULL;
 	}
-
-	dev_put(*dev);
 	return nl_rule;
 }

@@ -552,7 +547,7 @@ static int nss_nlipsec_op_add_sa(struct
 	sa_data->cmn.keys.auth_key = sa_rule->auth_key;
 	sa_data->cmn.keys.nonce = sa_rule->nonce;

-	error = nss_ipsecmgr_sa_add_sync(dev, &sa_rule->tuple, sa_data, &if_num);
+	error = nss_ipsecmgr_sa_add(dev, &sa_rule->tuple, sa_data, &if_num);
 	if (error) {
 		nss_nl_error("%d: Failed to add SA for net device(%s), error:%d\n", pid, nl_rule->ifname, error);
 		goto free_dev;
@@ -630,72 +625,73 @@ static int nss_nlipsec_op_delete_sa(stru
  * nss_nlipsec_op_add_flow()
  *	Add a flow
  */
-static int nss_nlipsec_op_add_flow(struct sk_buff *skb, struct genl_info *info)
-{
-	struct nss_ipsecmgr_flow_tuple *flow_tuple;
-	struct nss_ipsecmgr_sa_tuple *sa_tuple;
-	struct nss_nlipsec_rule *nl_rule;
-	struct net_device *dev;
-	uint32_t pid;
-	int error = 0;
-
-	nl_rule = nss_nlipsec_get_rule(info, NSS_NLIPSEC_CMD_ADD_FLOW, &dev);
-	if (!nl_rule) {
-		nss_nl_error("Failed to extract SA data\n");
-		return -EINVAL;
-	}
-
-	pid = nl_rule->cm.pid;
-	nss_nl_error("%d: device(%s)", pid, dev->name);
-
-	flow_tuple = &nl_rule->rule.flow.tuple;
-	sa_tuple = &nl_rule->rule.flow.sa;
-
-	error = nss_ipsecmgr_flow_add_sync(dev, flow_tuple, sa_tuple);
-	if (error) {
-		nss_nl_error("%d: Failed to add subnet for net_device(%s)", pid, nl_rule->ifname);
-	}
-
-	/*
-	 *  dev_put for dev_get done on nss_nlipsec_get_rule
-	 */
-	dev_put(dev);
-	return error;
-}
+// static int nss_nlipsec_op_add_flow(struct sk_buff *skb, struct genl_info *info)
+// {
+// 	struct nss_ipsecmgr_flow_tuple *flow_tuple;
+// 	struct nss_ipsecmgr_sa_tuple *sa_tuple;
+// 	struct nss_nlipsec_rule *nl_rule;
+// 	struct net_device *dev;
+// 	uint32_t pid;
+// 	int error = 0;
+//
+// 	nl_rule = nss_nlipsec_get_rule(info, NSS_NLIPSEC_CMD_ADD_FLOW, &dev);
+// 	if (!nl_rule) {
+// 		nss_nl_error("Failed to extract SA data\n");
+// 		return -EINVAL;
+// 	}
+//
+// 	pid = nl_rule->cm.pid;
+// 	nss_nl_error("%d: device(%s)", pid, dev->name);
+//
+// 	flow_tuple = &nl_rule->rule.flow.tuple;
+// 	sa_tuple = &nl_rule->rule.flow.sa;
+//
+// 	//struct nss_ipsecmgr_ref *nss_ipsecmgr_flow_alloc(struct nss_ipsecmgr_priv *priv, struct nss_ipsecmgr_key *key)
+// 	error = nss_ipsecmgr_flow_add_sync(dev, flow_tuple, sa_tuple);
+// 	if (error) {
+// 		nss_nl_error("%d: Failed to add subnet for net_device(%s)", pid, nl_rule->ifname);
+// 	}
+//
+// 	/*
+// 	 *  dev_put for dev_get done on nss_nlipsec_get_rule
+// 	 */
+// 	dev_put(dev);
+// 	return error;
+// }

 /*
  * nss_nlipsec_op_delete_flow()
  *	Delete a flow
  */
-static int nss_nlipsec_op_delete_flow(struct sk_buff *skb, struct genl_info *info)
-{
-	struct nss_ipsecmgr_flow_tuple *flow_tuple;
-	struct nss_ipsecmgr_sa_tuple *sa_tuple;
-	struct nss_nlipsec_rule *nl_rule;
-	struct net_device *dev;
-	uint32_t pid;
-	int error = 0;
-
-	nl_rule = nss_nlipsec_get_rule(info, NSS_NLIPSEC_CMD_DEL_FLOW, &dev);
-	if (!nl_rule) {
-		nss_nl_error("Failed to extract SA data\n");
-		return -EINVAL;
-	}
-
-	pid = nl_rule->cm.pid;
-	nss_nl_error("%d: device(%s)", pid, dev->name);
-
-	flow_tuple = &nl_rule->rule.flow.tuple;
-	sa_tuple = &nl_rule->rule.flow.sa;
-
-	nss_ipsecmgr_flow_del(dev, flow_tuple, sa_tuple);
-
-	/*
-	 *  dev_put for dev_get done on nss_nlipsec_get_rule
-	 */
-	dev_put(dev);
-	return error;
-}
+// static int nss_nlipsec_op_delete_flow(struct sk_buff *skb, struct genl_info *info)
+// {
+// 	struct nss_ipsecmgr_flow_tuple *flow_tuple;
+// 	struct nss_ipsecmgr_sa_tuple *sa_tuple;
+// 	struct nss_nlipsec_rule *nl_rule;
+// 	struct net_device *dev;
+// 	uint32_t pid;
+// 	int error = 0;
+//
+// 	nl_rule = nss_nlipsec_get_rule(info, NSS_NLIPSEC_CMD_DEL_FLOW, &dev);
+// 	if (!nl_rule) {
+// 		nss_nl_error("Failed to extract SA data\n");
+// 		return -EINVAL;
+// 	}
+//
+// 	pid = nl_rule->cm.pid;
+// 	nss_nl_error("%d: device(%s)", pid, dev->name);
+//
+// 	flow_tuple = &nl_rule->rule.flow.tuple;
+// 	sa_tuple = &nl_rule->rule.flow.sa;
+//
+// 	nss_ipsecmgr_flow_del(dev, flow_tuple, sa_tuple);
+//
+// 	/*
+// 	 *  dev_put for dev_get done on nss_nlipsec_get_rule
+// 	 */
+// 	dev_put(dev);
+// 	return error;
+// }

 /*
  * nss_nlipsec_init()
--- a/netlink/nss_nlipv4.c
+++ b/netlink/nss_nlipv4.c
@@ -71,20 +71,6 @@ static int nss_nlipv4_ops_destroy_rule(s
 static int nss_nlipv4_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * IPv4 family definition
- */
-static struct genl_family nss_nlipv4_family = {
-	.id = GENL_ID_GENERATE,				/* Auto generate ID */
-	.name = NSS_NLIPV4_FAMILY,			/* family name string */
-	.hdrsize = sizeof(struct nss_nlipv4_rule),	/* NSS NETLINK IPv4 rule */
-	.version = NSS_NL_VER,				/* Set it to NSS_NLIPv4 version */
-	.maxattr = NSS_IPV4_MAX_MSG_TYPES,		/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlipv4_mcgrp[] = {
@@ -100,6 +86,26 @@ static struct genl_ops nss_nlipv4_ops[]
 };

 /*
+ * IPv4 family definition
+ */
+static struct genl_family nss_nlipv4_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,				/* Auto generate ID */
+#endif
+	.name = NSS_NLIPV4_FAMILY,			/* family name string */
+	.hdrsize = sizeof(struct nss_nlipv4_rule),	/* NSS NETLINK IPv4 rule */
+	.version = NSS_NL_VER,				/* Set it to NSS_NLIPv4 version */
+	.maxattr = NSS_IPV4_MAX_MSG_TYPES,		/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlipv4_ops,
+	.n_ops = ARRAY_SIZE(nss_nlipv4_ops),
+	.mcgrps = nss_nlipv4_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlipv4_mcgrp)
+};
+
+/*
  * statistics call back handler for ipv4 from NSS
  */
 static struct notifier_block nss_ipv4_stats_notifier_nb = {
@@ -915,7 +921,7 @@ bool nss_nlipv4_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlipv4_family, nss_nlipv4_ops, nss_nlipv4_mcgrp);
+	error = genl_register_family(&nss_nlipv4_family);
 	if (error != 0) {
 		nss_nl_info_always("Error: unable to register IPv4 family\n");
 		return false;
--- a/netlink/nss_nlipv4_reasm.c
+++ b/netlink/nss_nlipv4_reasm.c
@@ -46,20 +46,6 @@ static int nss_nlipv4_reasm_ops_get_stat
 static int nss_nlipv4_reasm_process_notify(struct notifier_block *nb,  unsigned long val, void *data);

 /*
- * ipv4_reasm family definition
- */
-static struct genl_family nss_nlipv4_reasm_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLIPV4_REASM_FAMILY,				/* family name string */
-	.hdrsize = sizeof(struct nss_ipv4_reasm_stats_notification),	/* NSS NETLINK ipv4_reasm stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLIPV4_REASM version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlipv4_reasm_mcgrp[] = {
@@ -74,6 +60,26 @@ static struct genl_ops nss_nlipv4_reasm_
 };

 /*
+ * ipv4_reasm family definition
+ */
+static struct genl_family nss_nlipv4_reasm_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLIPV4_REASM_FAMILY,				/* family name string */
+	.hdrsize = sizeof(struct nss_ipv4_reasm_stats_notification),	/* NSS NETLINK ipv4_reasm stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLIPV4_REASM version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlipv4_reasm_ops,
+	.n_ops = ARRAY_SIZE(nss_nlipv4_reasm_ops),
+	.mcgrps = nss_nlipv4_reasm_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlipv4_reasm_mcgrp)
+};
+
+/*
  * stats call back handler for ipv4_reasm from NSS
  */
 static struct notifier_block nss_ipv4_reasm_stats_notifier_nb = {
@@ -125,7 +131,7 @@ bool nss_nlipv4_reasm_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlipv4_reasm_family, nss_nlipv4_reasm_ops, nss_nlipv4_reasm_mcgrp);
+	error = genl_register_family(&nss_nlipv4_reasm_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register ipv4_reasm family\n");
 		return false;
--- a/netlink/nss_nlipv6.c
+++ b/netlink/nss_nlipv6.c
@@ -79,20 +79,6 @@ static int nss_nlipv6_ops_destroy_rule(s
 static int nss_nlipv6_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * IPV6 family definition
- */
-static struct genl_family nss_nlipv6_family = {
-	.id = GENL_ID_GENERATE,				/* Auto generate ID */
-	.name = NSS_NLIPV6_FAMILY,			/* family name string */
-	.hdrsize = sizeof(struct nss_nlipv6_rule),	/* NSS NETLINK IPV6 rule */
-	.version = NSS_NL_VER,				/* Set it to NSS_NLIPV6 version */
-	.maxattr = NSS_IPV6_MAX_MSG_TYPES,		/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static struct genl_multicast_group nss_nlipv6_mcgrp[] = {
@@ -108,6 +94,26 @@ static struct genl_ops nss_nlipv6_ops[]
 };

 /*
+ * IPV6 family definition
+ */
+static struct genl_family nss_nlipv6_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,				/* Auto generate ID */
+#endif
+	.name = NSS_NLIPV6_FAMILY,			/* family name string */
+	.hdrsize = sizeof(struct nss_nlipv6_rule),	/* NSS NETLINK IPV6 rule */
+	.version = NSS_NL_VER,				/* Set it to NSS_NLIPV6 version */
+	.maxattr = NSS_IPV6_MAX_MSG_TYPES,		/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlipv6_ops,
+	.n_ops = ARRAY_SIZE(nss_nlipv6_ops),
+	.mcgrps = nss_nlipv6_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlipv6_mcgrp)
+};
+
+/*
  * statistics call back handler for ipv6 from NSS
  */
 static struct notifier_block nss_ipv6_stats_notifier_nb = {
@@ -131,7 +137,11 @@ static struct neighbour *nss_nlipv6_get_

 	IPV6_ADDR_TO_IN6_ADDR(daddr, dst_addr);

+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0))
 	rt = rt6_lookup(&init_net, &daddr, NULL, 0, 0);
+#else
+	rt = rt6_lookup(&init_net, &daddr, NULL, 0, NULL, 0);
+#endif
 	if (!rt) {
 		return NULL;
 	}
@@ -916,7 +926,7 @@ bool nss_nlipv6_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlipv6_family, nss_nlipv6_ops, nss_nlipv6_mcgrp);
+	error = genl_register_family(&nss_nlipv6_family);
 	if (error != 0) {
 		nss_nl_info_always("Error: unable to register IPV6 family\n");
 		return false;
--- a/netlink/nss_nlipv6_reasm.c
+++ b/netlink/nss_nlipv6_reasm.c
@@ -46,20 +46,6 @@ static int nss_nlipv6_reasm_ops_get_stat
 static int nss_nlipv6_reasm_process_notify(struct notifier_block *nb,  unsigned long val, void *data);

 /*
- * ipv6_reasm family definition
- */
-static struct genl_family nss_nlipv6_reasm_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLIPV6_REASM_FAMILY,				/* family name string */
-	.hdrsize = sizeof(struct nss_ipv6_reasm_stats_notification),	/* NSS NETLINK ipv6_reasm stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLIPV6_REASM version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlipv6_reasm_mcgrp[] = {
@@ -74,6 +60,26 @@ static struct genl_ops nss_nlipv6_reasm_
 };

 /*
+ * ipv6_reasm family definition
+ */
+static struct genl_family nss_nlipv6_reasm_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLIPV6_REASM_FAMILY,				/* family name string */
+	.hdrsize = sizeof(struct nss_ipv6_reasm_stats_notification),	/* NSS NETLINK ipv6_reasm stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLIPV6_REASM version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlipv6_reasm_ops,
+	.n_ops = ARRAY_SIZE(nss_nlipv6_reasm_ops),
+	.mcgrps = nss_nlipv6_reasm_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlipv6_reasm_mcgrp)
+};
+
+/*
  * stats call back handler for ipv6_reasm from NSS
  */
 static struct notifier_block nss_ipv6_reasm_stats_notifier_nb = {
@@ -125,7 +131,7 @@ bool nss_nlipv6_reasm_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlipv6_reasm_family, nss_nlipv6_reasm_ops, nss_nlipv6_reasm_mcgrp);
+	error = genl_register_family(&nss_nlipv6_reasm_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register ipv6_reasm family\n");
 		return false;
--- a/netlink/nss_nll2tpv2.c
+++ b/netlink/nss_nll2tpv2.c
@@ -47,20 +47,6 @@ static int nss_nll2tpv2_ops_get_stats(st
 static int nss_nll2tpv2_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * l2tpv2 family definition
- */
-static struct genl_family nss_nll2tpv2_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLL2TPV2_FAMILY,					/* family name string */
-	.hdrsize = sizeof(struct nss_l2tpv2_stats_notification),	/* NSS NETLINK l2tpv2 stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLL2TPV2 version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nll2tpv2_mcgrp[] = {
@@ -75,6 +61,26 @@ static struct genl_ops nss_nll2tpv2_ops[
 };

 /*
+ * l2tpv2 family definition
+ */
+static struct genl_family nss_nll2tpv2_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLL2TPV2_FAMILY,					/* family name string */
+	.hdrsize = sizeof(struct nss_l2tpv2_stats_notification),	/* NSS NETLINK l2tpv2 stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLL2TPV2 version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nll2tpv2_ops,
+	.n_ops = ARRAY_SIZE(nss_nll2tpv2_ops),
+	.mcgrps = nss_nll2tpv2_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nll2tpv2_mcgrp)
+};
+
+/*
  * device call back handler for l2tpv2 from NSS
  */
 static struct notifier_block nss_l2tpv2_stats_notifier_nb = {
@@ -126,7 +132,7 @@ bool nss_nll2tpv2_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nll2tpv2_family, nss_nll2tpv2_ops, nss_nll2tpv2_mcgrp);
+	error = genl_register_family(&nss_nll2tpv2_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register l2tpv2 family\n");
 		return false;
--- a/netlink/nss_nllso_rx.c
+++ b/netlink/nss_nllso_rx.c
@@ -47,20 +47,6 @@ static int nss_nllso_rx_ops_get_stats(st
 static int nss_nllso_rx_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * lso_rx family definition
- */
-static struct genl_family nss_nllso_rx_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLLSO_RX_FAMILY,					/* family name string */
-	.hdrsize = sizeof(struct nss_lso_rx_stats_notification),	/* NSS NETLINK lso_rx stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLLSO_RX version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nllso_rx_mcgrp[] = {
@@ -75,6 +61,26 @@ static struct genl_ops nss_nllso_rx_ops[
 };

 /*
+ * lso_rx family definition
+ */
+static struct genl_family nss_nllso_rx_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLLSO_RX_FAMILY,					/* family name string */
+	.hdrsize = sizeof(struct nss_lso_rx_stats_notification),	/* NSS NETLINK lso_rx stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLLSO_RX version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nllso_rx_ops,
+	.n_ops = ARRAY_SIZE(nss_nllso_rx_ops),
+	.mcgrps = nss_nllso_rx_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nllso_rx_mcgrp)
+};
+
+/*
  * device call back handler for lso_rx from NSS
  */
 static struct notifier_block nss_lso_rx_stats_notifier_nb = {
@@ -126,7 +132,7 @@ bool nss_nllso_rx_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nllso_rx_family, nss_nllso_rx_ops, nss_nllso_rx_mcgrp);
+	error = genl_register_family(&nss_nllso_rx_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register lso_rx family\n");
 		return false;
--- a/netlink/nss_nlmap_t.c
+++ b/netlink/nss_nlmap_t.c
@@ -47,20 +47,6 @@ static int nss_nlmap_t_ops_get_stats(str
 static int nss_nlmap_t_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * map_t family definition
- */
-static struct genl_family nss_nlmap_t_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLMAP_T_FAMILY,					/* family name string */
-	.hdrsize = sizeof(struct nss_map_t_stats_notification),		/* NSS NETLINK map_t stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLMAP_T version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlmap_t_mcgrp[] = {
@@ -75,6 +61,26 @@ static struct genl_ops nss_nlmap_t_ops[]
 };

 /*
+ * map_t family definition
+ */
+static struct genl_family nss_nlmap_t_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLMAP_T_FAMILY,					/* family name string */
+	.hdrsize = sizeof(struct nss_map_t_stats_notification),		/* NSS NETLINK map_t stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLMAP_T version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlmap_t_ops,
+	.n_ops = ARRAY_SIZE(nss_nlmap_t_ops),
+	.mcgrps = nss_nlmap_t_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlmap_t_mcgrp)
+};
+
+/*
  * device call back handler for map_t from NSS
  */
 static struct notifier_block nss_map_t_stats_notifier_nb = {
@@ -126,7 +132,7 @@ bool nss_nlmap_t_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlmap_t_family, nss_nlmap_t_ops, nss_nlmap_t_mcgrp);
+	error = genl_register_family(&nss_nlmap_t_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register map_t family\n");
 		return false;
--- a/netlink/nss_nln2h.c
+++ b/netlink/nss_nln2h.c
@@ -46,20 +46,6 @@ static int nss_nln2h_ops_get_stats(struc
 static int nss_nln2h_process_notify(struct notifier_block *nb,  unsigned long val, void *data);

 /*
- * n2h family definition
- */
-static struct genl_family nss_nln2h_family = {
-	.id = GENL_ID_GENERATE,					/* Auto generate ID */
-	.name = NSS_NLN2H_FAMILY,				/* family name string */
-	.hdrsize = sizeof(struct nss_n2h_stats_notification),	/* NSS NETLINK n2h stats */
-	.version = NSS_NL_VER,					/* Set it to NSS_NLN2H version */
-	.maxattr = NSS_STATS_EVENT_MAX,				/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nln2h_mcgrp[] = {
@@ -74,6 +60,26 @@ static struct genl_ops nss_nln2h_ops[] =
 };

 /*
+ * n2h family definition
+ */
+static struct genl_family nss_nln2h_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,					/* Auto generate ID */
+#endif
+	.name = NSS_NLN2H_FAMILY,				/* family name string */
+	.hdrsize = sizeof(struct nss_n2h_stats_notification),	/* NSS NETLINK n2h stats */
+	.version = NSS_NL_VER,					/* Set it to NSS_NLN2H version */
+	.maxattr = NSS_STATS_EVENT_MAX,				/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nln2h_ops,
+	.n_ops = ARRAY_SIZE(nss_nln2h_ops),
+	.mcgrps = nss_nln2h_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nln2h_mcgrp)
+};
+
+/*
  * stats call back handler for n2h from NSS
  */
 static struct notifier_block nss_n2h_stats_notifier_nb = {
@@ -125,7 +131,7 @@ bool nss_nln2h_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nln2h_family, nss_nln2h_ops, nss_nln2h_mcgrp);
+	error = genl_register_family(&nss_nln2h_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register n2h family\n");
 		return false;
--- a/netlink/nss_nloam.c
+++ b/netlink/nss_nloam.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2016,2018, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2016,2018, 2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -82,11 +82,25 @@ struct nss_nloam_msg_map global_nloam_ms
 	},
 };

+static struct genl_multicast_group nss_nloam_mcgrp[] = {
+	{.name = NSS_NLOAM_MCAST_GRP},
+};
+
+/*
+ * operation table called by the generic netlink layer based on the command
+ */
+static struct genl_ops nss_nloam_ops[] = {
+	{.cmd = NSS_NLOAM_CMD_NONE, .doit = nss_nloam_op_none,},
+	{.cmd = NSS_NLOAM_CMD_GET_REQ, .doit = nss_nloam_op_get_req,},
+};
+
 /*
  * OAM family definition
  */
 static struct genl_family nss_nloam_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
 	.id = GENL_ID_GENERATE,				/* Auto generate ID */
+#endif
 	.name = NSS_NLOAM_FAMILY,			/* family name string */
 	.hdrsize = sizeof(struct nss_nloam_rule),	/* NLOAM msg size */
 	.version = NSS_NL_VER,				/* nss netlink version */
@@ -94,18 +108,10 @@ static struct genl_family nss_nloam_fami
 	.netnsok = false,
 	.pre_doit = NULL,
 	.post_doit = NULL,
-};
-
-static struct genl_multicast_group nss_nloam_mcgrp[] = {
-	{.name = NSS_NLOAM_MCAST_GRP},
-};
-
-/*
- * operation table called by the generic netlink layer based on the command
- */
-static struct genl_ops nss_nloam_ops[] = {
-	{.cmd = NSS_NLOAM_CMD_NONE, .doit = nss_nloam_op_none,},
-	{.cmd = NSS_NLOAM_CMD_GET_REQ, .doit = nss_nloam_op_get_req,},
+	.ops = nss_nloam_ops,
+	.n_ops = ARRAY_SIZE(nss_nloam_ops),
+	.mcgrps = nss_nloam_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nloam_mcgrp)
 };

 #define NSS_NLOAM_OPS_SZ ARRAY_SIZE(nss_nloam_ops)
@@ -340,7 +346,7 @@ bool nss_nloam_init(void)
 	/*
 	 * register with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nloam_family, nss_nloam_ops, nss_nloam_mcgrp);
+	error = genl_register_family(&nss_nloam_family);
 	if (error != 0) {
 		nss_nl_info_always("unable to register OAM family\n");
 		return false;
--- a/netlink/nss_nlpppoe.c
+++ b/netlink/nss_nlpppoe.c
@@ -47,20 +47,6 @@ static int nss_nlpppoe_ops_get_stats(str
 static int nss_nlpppoe_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * pppoe family definition
- */
-static struct genl_family nss_nlpppoe_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLPPPOE_FAMILY,					/* family name string */
-	.hdrsize = sizeof(struct nss_pppoe_stats_notification),		/* NSS NETLINK pppoe stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLPPPOE version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlpppoe_mcgrp[] = {
@@ -75,6 +61,26 @@ static struct genl_ops nss_nlpppoe_ops[]
 };

 /*
+ * pppoe family definition
+ */
+static struct genl_family nss_nlpppoe_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLPPPOE_FAMILY,					/* family name string */
+	.hdrsize = sizeof(struct nss_pppoe_stats_notification),		/* NSS NETLINK pppoe stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLPPPOE version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlpppoe_ops,
+	.n_ops = ARRAY_SIZE(nss_nlpppoe_ops),
+	.mcgrps = nss_nlpppoe_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlpppoe_mcgrp)
+};
+
+/*
  * device call back handler for pppoe from NSS
  */
 static struct notifier_block nss_pppoe_stats_notifier_nb = {
@@ -126,7 +132,7 @@ bool nss_nlpppoe_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlpppoe_family, nss_nlpppoe_ops, nss_nlpppoe_mcgrp);
+	error = genl_register_family(&nss_nlpppoe_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register pppoe family\n");
 		return false;
--- a/netlink/nss_nlpptp.c
+++ b/netlink/nss_nlpptp.c
@@ -47,20 +47,6 @@ static int nss_nlpptp_ops_get_stats(stru
 static int nss_nlpptp_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * pptp family definition
- */
-static struct genl_family nss_nlpptp_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLPPTP_FAMILY,					/* family name string */
-	.hdrsize = sizeof(struct nss_pptp_stats_notification),		/* NSS NETLINK pptp stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLPPTP version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlpptp_mcgrp[] = {
@@ -75,6 +61,26 @@ static struct genl_ops nss_nlpptp_ops[]
 };

 /*
+ * pptp family definition
+ */
+static struct genl_family nss_nlpptp_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLPPTP_FAMILY,					/* family name string */
+	.hdrsize = sizeof(struct nss_pptp_stats_notification),		/* NSS NETLINK pptp stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLPPTP version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlpptp_ops,
+	.n_ops = ARRAY_SIZE(nss_nlpptp_ops),
+	.mcgrps = nss_nlpptp_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlpptp_mcgrp)
+};
+
+/*
  * device call back handler for pptp from NSS
  */
 static struct notifier_block nss_pptp_stats_notifier_nb = {
@@ -126,7 +132,7 @@ bool nss_nlpptp_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlpptp_family, nss_nlpptp_ops, nss_nlpptp_mcgrp);
+	error = genl_register_family(&nss_nlpptp_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register pptp family\n");
 		return false;
--- a/netlink/nss_nlwifili.c
+++ b/netlink/nss_nlwifili.c
@@ -47,20 +47,6 @@ static int nss_nlwifili_ops_get_stats(st
 static int nss_nlwifili_process_notify(struct notifier_block *nb, unsigned long val, void *data);

 /*
- * wifili family definition
- */
-static struct genl_family nss_nlwifili_family = {
-	.id = GENL_ID_GENERATE,						/* Auto generate ID */
-	.name = NSS_NLWIFILI_FAMILY,					/* family name string */
-	.hdrsize = sizeof(struct nss_wifili_stats_notification),	/* NSS NETLINK wifili stats */
-	.version = NSS_NL_VER,						/* Set it to NSS_NLWIFILI version */
-	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
-	.netnsok = true,
-	.pre_doit = NULL,
-	.post_doit = NULL,
-};
-
-/*
  * multicast group for sending message status & events
  */
 static const struct genl_multicast_group nss_nlwifili_mcgrp[] = {
@@ -75,6 +61,26 @@ static struct genl_ops nss_nlwifili_ops[
 };

 /*
+ * wifili family definition
+ */
+static struct genl_family nss_nlwifili_family = {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 9, 0))
+	.id = GENL_ID_GENERATE,						/* Auto generate ID */
+#endif
+	.name = NSS_NLWIFILI_FAMILY,					/* family name string */
+	.hdrsize = sizeof(struct nss_wifili_stats_notification),	/* NSS NETLINK wifili stats */
+	.version = NSS_NL_VER,						/* Set it to NSS_NLWIFILI version */
+	.maxattr = NSS_STATS_EVENT_MAX,					/* maximum commands supported */
+	.netnsok = true,
+	.pre_doit = NULL,
+	.post_doit = NULL,
+	.ops = nss_nlwifili_ops,
+	.n_ops = ARRAY_SIZE(nss_nlwifili_ops),
+	.mcgrps = nss_nlwifili_mcgrp,
+	.n_mcgrps = ARRAY_SIZE(nss_nlwifili_mcgrp)
+};
+
+/*
  * device call back handler for wifili from NSS
  */
 static struct notifier_block nss_wifili_stats_notifier_nb = {
@@ -126,7 +132,7 @@ bool nss_nlwifili_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	error = genl_register_family_with_ops_groups(&nss_nlwifili_family, nss_nlwifili_ops, nss_nlwifili_mcgrp);
+	error = genl_register_family(&nss_nlwifili_family);
 	if (error) {
 		nss_nl_info_always("Error: unable to register wifili family\n");
 		return false;
--- a/netlink/nss_nlcapwap.c
+++ b/netlink/nss_nlcapwap.c
@@ -36,7 +36,6 @@
 #include <nss_dtls_cmn.h>
 #include <nss_nlcmn_if.h>
 #include <nss_nl_if.h>
-#include "nss_crypto_defines.h"
 #include "nss_nl.h"
 #include "nss_nlcapwap_if.h"
 #include "nss_nlcapwap.h"
@@ -1510,7 +1509,7 @@ bool nss_nlcapwap_init(void)
 	/*
 	 * register NETLINK ops with the family
 	 */
-	err = genl_register_family_with_ops_groups(&nss_nlcapwap_family, nss_nlcapwap_cmd_ops, nss_nlcapwap_family_mcgrp);
+	err = genl_register_family(&nss_nlcapwap_family);
 	if (err) {
 		nss_nl_info_always("Error: %d unable to register capwap family\n", err);
 		goto free;
