--- a/nss_qdisc/Makefile
+++ b/nss_qdisc/Makefile
@@ -9,18 +9,16 @@ endif
 ccflags-y += -Wall -Werror

 obj-m += qca-nss-qdisc.o
-qca-nss-qdisc-objs := 	nss_bf.o \
-			nss_blackhole.o \
-			nss_codel.o \
+qca-nss-qdisc-objs := 	nss_qdisc.o \
 			nss_fifo.o \
-			nss_htb.o \
-			nss_prio.o \
-			nss_qdisc.o \
-			nss_qdisc_htable.o \
-			nss_qdisc_stats.o \
+			nss_codel.o \
 			nss_tbl.o \
-			nss_wred.o \
-			nss_wrr.o
+			nss_prio.o \
+			nss_bf.o \
+			nss_wrr.o \
+			nss_htb.o \
+			nss_blackhole.o \
+			nss_wred.o

 ifeq ($(SoC),$(filter $(SoC),ipq807x ipq807x_64 ipq60xx ipq60xx_64))
 qca-nss-qdisc-objs += 	nss_ppe.o \
--- a/nss_qdisc/igs/Makefile
+++ b/nss_qdisc/igs/Makefile
@@ -1,7 +1,6 @@
 # Makefile for IGS (Ingress Shaping)

 ccflags-y += $(NSS_CCFLAGS) -I$(obj)/../../exports
-ccflags-y += -DNSS_IGS_DEBUG_LEVEL=2
 ccflags-y += -Wall -Werror

 obj-m += act_nssmirred.o
--- a/nss_qdisc/igs/nss_mirred.c
+++ b/nss_qdisc/igs/nss_mirred.c
@@ -82,20 +82,24 @@ static const struct nla_policy nss_mirre
  * nss_mirred_init()
  *	Initialize the nss mirred action.
  */
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0))
 static int nss_mirred_init(struct net *net, struct nlattr *nla,
-			   struct nlattr *est, struct tc_action *tc_act, int ovr,
-			   int bind)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0))
+						struct nlattr *est, struct tc_action *tc_act, int ovr,
+						int bind)
+{
+#elif (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0))
+						struct nlattr *est, struct tc_action **tc_act, int ovr,
+						int bind, bool rtnl_held, struct tcf_proto *tp,
+						u32 flags, struct netlink_ext_ack *extack)
 {
 #else
-static int nss_mirred_init(struct net *net, struct nlattr *nla,
-			   struct nlattr *est, struct tc_action **tc_act, int ovr,
-			   int bind, bool rtnl_held, struct tcf_proto *tp,
-			   struct netlink_ext_ack *extack)
+						struct nlattr *est, struct tc_action **tc_act,
+						struct tcf_proto *tp, u32 flags, struct netlink_ext_ack *extack)
 {
+	bool bind = flags & TCA_ACT_FLAGS_BIND;
+#endif
 	struct tc_action_net *tn = net_generic(net, nss_mirred_net_id);
 	u32 index;
-#endif
 	struct nlattr *arr[TC_NSS_MIRRED_MAX + 1];
 	struct tc_nss_mirred *parm;
 	struct nss_mirred_tcf *act;
@@ -239,8 +243,13 @@ static int nss_mirred_init(struct net *n
 	}

 	if (!ret) {
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 5, 0))
 		ret = tcf_idr_create(tn, index, est, tc_act, &nss_mirred_act_ops,
 				bind, true);
+#else
+		ret = tcf_idr_create(tn, index, est, tc_act, &nss_mirred_act_ops,
+				bind, true, 0);
+#endif
 		if (ret) {
 			tcf_idr_cleanup(tn, index);
 			return ret;
--- a/nss_qdisc/nss_bf.c
+++ b/nss_qdisc/nss_bf.c
@@ -1,13 +1,9 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2017, 2019-2021, The Linux Foundation. All rights reserved.
- *
- * Copyright (c) 2022, Qualcomm Innovation Center, Inc. All rights reserved.
- *
+ * Copyright (c) 2014-2017, 2019-2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
- *
  * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
@@ -178,6 +174,11 @@ static int nss_bf_change_class(struct Qd
 		 */
 		qdisc_class_hash_grow(sch, &q->clhash);

+		/*
+		 * Start the stats polling timer
+		 */
+		nss_qdisc_start_basic_stats_polling(&cl->nq);
+
 		nss_qdisc_info("Class %u successfully allocated\n", classid);
 	}

@@ -276,6 +277,11 @@ static void nss_bf_destroy_class(struct
 	 nss_qdisc_put(cl->qdisc);

 	/*
+	 * Stop the stats polling timer and free class
+	 */
+	nss_qdisc_stop_basic_stats_polling(&cl->nq);
+
+	/*
 	 * Destroy the shaper in NSS
 	 */
 	nss_qdisc_destroy(&cl->nq);
@@ -290,7 +296,11 @@ static void nss_bf_destroy_class(struct
  * nss_bf_delete_class()
  *	Detaches a class from operation, but does not destroy it.
  */
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0))
+static int nss_bf_delete_class(struct Qdisc *sch, unsigned long arg, struct netlink_ext_ack *extack)
+#else
 static int nss_bf_delete_class(struct Qdisc *sch, unsigned long arg)
+#endif
 {
 	struct nss_bf_sched_data *q = qdisc_priv(sch);
 	struct nss_bf_class_data *cl = (struct nss_bf_class_data *)arg;
@@ -299,14 +309,11 @@ static int nss_bf_delete_class(struct Qd
 	struct nss_qdisc *nq_child = (struct nss_qdisc *)qdisc_priv(cl->qdisc);

 	/*
-	 * If the class is the root class or has qdiscs attached, we do not
-	 * support deleting it.
+	 * Since all classes are leaf nodes in our case, we dont have to make
+	 * that check.
 	 */
-	if ((cl == &q->root) || (cl->qdisc != &noop_qdisc)) {
-		nss_qdisc_warning("Cannot delete bf class %x as it is the root "
-				  "class or has child qdisc attached\n", cl->nq.qos_tag);
+	if (cl == &q->root)
 		return -EBUSY;
-	}

 	/*
 	 * The message to NSS should be sent to the parent of this class
@@ -320,22 +327,14 @@ static int nss_bf_delete_class(struct Qd
 	}

 	sch_tree_lock(sch);
+	qdisc_reset(cl->qdisc);
 	qdisc_class_hash_remove(&q->clhash, &cl->cl_common);
 	refcnt = nss_qdisc_atomic_sub_return(&cl->nq);
 	sch_tree_unlock(sch);
-
-	/*
-	 * For 5.4 and above kernels, calling nss_htb_destroy_class
-	 * explicitly as there is no put_class which would have called
-	 * nss_bf_destroy_class when refcnt becomes zero.
-	 */
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0))
-	nss_bf_destroy_class(sch, cl);
-#else
 	if (!refcnt) {
 		nss_qdisc_error("Reference count should not be zero for class %px\n", cl);
 	}
-#endif
+
 	return 0;
 }

@@ -635,11 +634,6 @@ static int nss_bf_change_qdisc(struct Qd
  */
 static void nss_bf_reset_class(struct nss_bf_class_data *cl)
 {
-	if (cl->qdisc == &noop_qdisc) {
-		nss_qdisc_trace("Class %x has no child qdisc to reset\n", cl->nq.qos_tag);
-		return;
-	}
-
 	nss_qdisc_reset(cl->qdisc);
 	nss_qdisc_info("Nssbf class resetted %px\n", cl->qdisc);
 }
@@ -721,6 +715,11 @@ static void nss_bf_destroy_qdisc(struct
 	qdisc_class_hash_destroy(&q->clhash);

 	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(&q->nq);
+
+	/*
 	 * Now we can go ahead and destroy the qdisc.
 	 * Note: We dont have to detach ourself from our parent because this
 	 *	 will be taken care of by the graft call.
@@ -799,6 +798,11 @@ static int nss_bf_init_qdisc(struct Qdis
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(&q->nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_blackhole.c
+++ b/nss_qdisc/nss_blackhole.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014, 2016-2017, 2020-2021,  The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014, 2016-2017, 2020,  The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -90,6 +90,11 @@ static void nss_blackhole_destroy(struct
 {
 	struct nss_qdisc *nq = (struct nss_qdisc *)qdisc_priv(sch);

+	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(nq);
+
 	nss_qdisc_info("destroying qdisc %x\n", sch->handle);
 	nss_qdisc_destroy(nq);
 }
@@ -219,6 +224,11 @@ static int nss_blackhole_init(struct Qdi
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_codel.c
+++ b/nss_qdisc/nss_codel.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014, 2016-2018, 2020-2021,  The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014, 2016-2018, 2020,  The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -155,6 +155,10 @@ static void nss_codel_destroy(struct Qdi
 {
 	struct nss_codel_sched_data *q = qdisc_priv(sch);

+	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(&q->nq);
 	nss_qdisc_destroy(&q->nq);
 	nss_codel_flow_queues_free(q);
 	nss_qdisc_info("nss_codel destroyed");
@@ -451,6 +455,11 @@ static int nss_codel_init(struct Qdisc *
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_fifo.c
+++ b/nss_qdisc/nss_fifo.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014, 2016-2017, 2020-2021, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014, 2016-2017, 2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -66,6 +66,11 @@ static void nss_fifo_destroy(struct Qdis
 {
 	struct nss_qdisc *nq = (struct nss_qdisc *)qdisc_priv(sch);

+	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(nq);
+
 	nss_qdisc_destroy(nq);
 	nss_qdisc_info("nss_fifo destroyed");
 }
@@ -291,6 +296,11 @@ static int nss_fifo_init(struct Qdisc *s
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_htb.c
+++ b/nss_qdisc/nss_htb.c
@@ -1,13 +1,9 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2017, 2019-2021, The Linux Foundation. All rights reserved.
- *
- * Copyright (c) 2022, Qualcomm Innovation Center, Inc. All rights reserved.
- *
+ * Copyright (c) 2014-2017, 2019-2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
- *
  * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
@@ -413,6 +409,11 @@ static int nss_htb_change_class(struct Q
 		 */
 		qdisc_class_hash_grow(sch, &q->clhash);

+		/*
+		 * Start the stats polling timer
+		 */
+		nss_qdisc_start_basic_stats_polling(&cl->nq);
+
 		nss_qdisc_trace("class %x successfully allocated and initialized\n", classid);
 	}

@@ -502,6 +503,11 @@ static void nss_htb_destroy_class(struct
 	 nss_qdisc_put(cl->qdisc);

 	/*
+	 * Stop the stats polling timer and free class
+	 */
+	nss_qdisc_stop_basic_stats_polling(&cl->nq);
+
+	/*
 	 * Destroy the shaper in NSS
 	 */
 	nss_qdisc_destroy(&cl->nq);
@@ -516,7 +522,11 @@ static void nss_htb_destroy_class(struct
  * nss_htb_delete_class()
  *	Detaches a class from operation, but does not destroy it.
  */
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0))
+static int nss_htb_delete_class(struct Qdisc *sch, unsigned long arg, struct netlink_ext_ack *extack)
+#else
 static int nss_htb_delete_class(struct Qdisc *sch, unsigned long arg)
+#endif
 {
 	struct nss_htb_sched_data *q = qdisc_priv(sch);
 	struct nss_htb_class_data *cl = (struct nss_htb_class_data *)arg;
@@ -524,12 +534,10 @@ static int nss_htb_delete_class(struct Q
 	int refcnt;

 	/*
-	 * If the class still has child nodes or qdiscs, then we do not
+	 * If the class still has child nodes, then we do not
 	 * support deleting it.
 	 */
-	if ((cl->children) || (cl->qdisc != &noop_qdisc)) {
-		nss_qdisc_warning("Cannot delete htb class %x with child nodes "
-				  "or qdisc attached\n", cl->nq.qos_tag);
+	if (cl->children) {
 		return -EBUSY;
 	}

@@ -560,21 +568,16 @@ static int nss_htb_delete_class(struct Q
 	}

 	sch_tree_lock(sch);
+	qdisc_reset(cl->qdisc);
 	qdisc_class_hash_remove(&q->clhash, &cl->sch_common);

 	/*
 	 * If we are root class, we dont have to update our parent.
 	 * We simply deduct refcnt and return.
-	 * For 5.4 and above kernels, calling nss_htb_destroy_class
-	 * explicitly as there is no put_class which would have called
-	 * nss_htb_destroy_class when refcnt becomes zero.
 	 */
 	if (!cl->parent) {
 		refcnt = nss_qdisc_atomic_sub_return(&cl->nq);
 		sch_tree_unlock(sch);
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0))
-		nss_htb_destroy_class(sch, cl);
-#endif
 		return 0;
 	}

@@ -593,14 +596,6 @@ static int nss_htb_delete_class(struct Q
 	refcnt = nss_qdisc_atomic_sub_return(&cl->nq);
 	sch_tree_unlock(sch);

-	/*
-	 * For 5.4 and above kernels, calling nss_htb_destroy_class
-	 * explicitly as there is no put_class which would have called
-	 * nss_htb_destroy_class when refcnt becomes zero.
-	 */
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0))
-	nss_htb_destroy_class(sch, cl);
-#endif
 	return 0;
 }

@@ -903,11 +898,6 @@ static int nss_htb_change_qdisc(struct Q
  */
 static void nss_htb_reset_class(struct nss_htb_class_data *cl)
 {
-	if (cl->qdisc == &noop_qdisc) {
-		nss_qdisc_trace("Class %x has no child qdisc to reset\n", cl->nq.qos_tag);
-		return;
-	}
-
 	nss_qdisc_reset(cl->qdisc);
 	nss_qdisc_trace("htb class %x reset\n", cl->nq.qos_tag);
 }
@@ -991,6 +981,11 @@ static void nss_htb_destroy_qdisc(struct
 	qdisc_class_hash_destroy(&q->clhash);

 	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(&q->nq);
+
+	/*
 	 * Now we can go ahead and destroy the qdisc.
 	 * Note: We dont have to detach ourself from our parent because this
 	 *	 will be taken care of by the graft call.
@@ -1066,6 +1061,11 @@ static int nss_htb_init_qdisc(struct Qdi
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(&q->nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_ppe.c
+++ b/nss_qdisc/nss_ppe.c
@@ -1,8 +1,6 @@
 /*
  **************************************************************************
  * Copyright (c) 2017-2020, The Linux Foundation. All rights reserved.
- *
- * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -128,7 +126,7 @@ static struct nss_ppe_res *nss_ppe_res_e

 	spin_lock_bh(&ppe_port->lock);
 	for (i = max; i > 0; i--) {
-		res = kzalloc(sizeof(struct nss_ppe_res), GFP_ATOMIC);
+		res = kzalloc(sizeof(struct nss_ppe_res), GFP_KERNEL);
 		if (!res) {
 			nss_qdisc_error("Free queue list allocation failed for port %u\n", port);
 			goto fail;
--- a/nss_qdisc/nss_prio.c
+++ b/nss_qdisc/nss_prio.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2017, 2019-2021, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014-2017, 2019-2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -136,6 +136,11 @@ static void nss_prio_destroy(struct Qdis
 	}

 	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(&q->nq);
+
+	/*
 	 * Destroy the qdisc in NSS
 	 */
 	nss_qdisc_destroy(&q->nq);
@@ -279,6 +284,10 @@ static int nss_prio_init(struct Qdisc *s
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(&q->nq);
 	return 0;
 }

--- a/nss_qdisc/nss_qdisc.c
+++ b/nss_qdisc/nss_qdisc.c
@@ -29,6 +29,9 @@

 void *nss_qdisc_ctx;			/* Shaping context for nss_qdisc */

+#define NSS_QDISC_COMMAND_TIMEOUT (10*HZ) /* We set 10sec to be the command */
+					   /* timeout value for messages */
+
 /*
  * Defines related to root hash maintenance
  */
@@ -36,53 +39,6 @@ void *nss_qdisc_ctx;			/* Shaping contex
 #define NSS_QDISC_ROOT_HASH_MASK (NSS_QDISC_ROOT_HASH_SIZE - 1)

 /*
- * nss_qdisc_get_interface_msg()
- *	Returns the correct message that needs to be sent down to the NSS interface.
- */
-int nss_qdisc_get_interface_msg(bool is_bridge, uint32_t msg_type)
-{
-	/*
-	 * We re-assign the message based on whether this is for the I shaper
-	 * or the B shaper. The is_bridge flag tells if we are on a bridge interface.
-	 */
-	if (is_bridge) {
-		switch (msg_type) {
-		case NSS_QDISC_IF_SHAPER_ASSIGN:
-			return NSS_IF_BSHAPER_ASSIGN;
-		case NSS_QDISC_IF_SHAPER_UNASSIGN:
-			return NSS_IF_BSHAPER_UNASSIGN;
-		case NSS_QDISC_IF_SHAPER_CONFIG:
-			return NSS_IF_BSHAPER_CONFIG;
-		default:
-			nss_qdisc_info("Unknown message type for a bridge - type %d", msg_type);
-			return -1;
-		}
-	} else {
-		switch (msg_type) {
-		case NSS_QDISC_IF_SHAPER_ASSIGN:
-			return NSS_IF_ISHAPER_ASSIGN;
-		case NSS_QDISC_IF_SHAPER_UNASSIGN:
-			return NSS_IF_ISHAPER_UNASSIGN;
-		case NSS_QDISC_IF_SHAPER_CONFIG:
-			return NSS_IF_ISHAPER_CONFIG;
-		default:
-			nss_qdisc_info("Unknown message type for an interface - type %d", msg_type);
-			return -1;
-		}
-	}
-}
-
-/*
- * nss_qdisc_msg_init()
- *      Initialize the qdisc specific message
- */
-void nss_qdisc_msg_init(struct nss_if_msg *nim, uint16_t if_num, uint32_t msg_type, uint32_t len,
-				nss_if_msg_callback_t cb, void *app_data)
-{
-	nss_cmn_msg_init(&nim->cm, if_num, msg_type, len, (void *)cb, app_data);
-}
-
-/*
  * nss_qdisc_interface_is_virtual()
  *	Return true if it is redirect or bridge interface.
  */
@@ -165,6 +121,53 @@ static int nss_qdisc_ppe_init(struct Qdi
 #endif

 /*
+ * nss_qdisc_msg_init()
+ *      Initialize the qdisc specific message
+ */
+static void nss_qdisc_msg_init(struct nss_if_msg *nim, uint16_t if_num, uint32_t msg_type, uint32_t len,
+				nss_if_msg_callback_t cb, void *app_data)
+{
+	nss_cmn_msg_init(&nim->cm, if_num, msg_type, len, (void*)cb, app_data);
+}
+
+/*
+ * nss_qdisc_get_interface_msg()
+ *	Returns the correct message that needs to be sent down to the NSS interface.
+ */
+static inline int nss_qdisc_get_interface_msg(bool is_bridge, uint32_t msg_type)
+{
+	/*
+	 * We re-assign the message based on whether this is for the I shaper
+	 * or the B shaper. The is_bridge flag tells if we are on a bridge interface.
+	 */
+	if (is_bridge) {
+		switch(msg_type) {
+		case NSS_QDISC_IF_SHAPER_ASSIGN:
+			return NSS_IF_BSHAPER_ASSIGN;
+		case NSS_QDISC_IF_SHAPER_UNASSIGN:
+			return NSS_IF_BSHAPER_UNASSIGN;
+		case NSS_QDISC_IF_SHAPER_CONFIG:
+			return NSS_IF_BSHAPER_CONFIG;
+		default:
+			nss_qdisc_info("Unknown message type for a bridge - type %d", msg_type);
+			return -1;
+		}
+	} else {
+		switch(msg_type) {
+		case NSS_QDISC_IF_SHAPER_ASSIGN:
+			return NSS_IF_ISHAPER_ASSIGN;
+		case NSS_QDISC_IF_SHAPER_UNASSIGN:
+			return NSS_IF_ISHAPER_UNASSIGN;
+		case NSS_QDISC_IF_SHAPER_CONFIG:
+			return NSS_IF_ISHAPER_CONFIG;
+		default:
+			nss_qdisc_info("Unknown message type for an interface - type %d", msg_type);
+			return -1;
+		}
+	}
+}
+
+/*
  * nss_qdisc_attach_bshaper_callback()
  *	Call back funtion for bridge shaper attach to an interface.
  */
@@ -609,6 +612,7 @@ static void nss_qdisc_root_cleanup_free_
 	nss_qdisc_info("Root qdisc %px (type %d) free SUCCESS - response "
 			"type: %d\n", nq->qdisc, nq->type,
 			nim->msg.shaper_configure.config.response_type);
+
 	nss_qdisc_root_cleanup_shaper_unassign(nq);
 }

@@ -1162,15 +1166,8 @@ unsigned int nss_qdisc_drop(struct Qdisc
  */
 void nss_qdisc_reset(struct Qdisc *sch)
 {
-	struct nss_qdisc *nq;
-
-	if(!(sch->flags & TCQ_F_NSS)) {
-		qdisc_reset_queue(sch);
-		nss_qdisc_info("Qdisc %px resetting non NSS qdisc\n", sch);
-		return;
-	}
+	struct nss_qdisc *nq = qdisc_priv(sch);

-	nq = qdisc_priv(sch);
 	nss_qdisc_info("Qdisc %px (type %d) resetting\n",
 			sch, nq->type);

@@ -1209,10 +1206,10 @@ static bool nss_qdisc_iterate_fl(struct
 		return 0;
 	}

-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0))
-	status = tc_classify(skb, tcf, &res, false);
-#else
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0))
 	status = tcf_classify(skb, tcf, &res, false);
+#else
+	status = tcf_classify(skb, NULL, tcf, &res, false);
 #endif
 	if ((status == TC_ACT_STOLEN) || (status == TC_ACT_QUEUED)) {
 		return 1;
@@ -1892,7 +1889,6 @@ int nss_qdisc_configure(struct nss_qdisc
 	return 0;
 }

-
 /*
  * nss_qdisc_register_configure_callback()
  *	Register shaper configure callback, which gets invoked on receiving a response.
@@ -1949,11 +1945,6 @@ void nss_qdisc_destroy(struct nss_qdisc
 	}

 	/*
-	 * Delete node from stats list
-	 */
-	nss_qdisc_stats_qdisc_detach(nq);
-
-	/*
 	 * How we begin to tidy up depends on whether we are root or child
 	 */
 	nq->pending_final_state = NSS_QDISC_STATE_IDLE;
@@ -1972,11 +1963,6 @@ void nss_qdisc_destroy(struct nss_qdisc
 		}

 		/*
-		 * Stop stats polling
-		 */
-		nss_qdisc_stats_stop_polling(nq);
-
-		/*
 		 * Begin by freeing the root shaper node
 		 */
 		nss_qdisc_root_cleanup_free_node(nq);
@@ -2129,8 +2115,6 @@ int __nss_qdisc_init(struct Qdisc *sch,
 	 */
 	if ((sch->parent == TC_H_ROOT) && (!nq->is_class)) {
 		nss_qdisc_info("Qdisc %px (type %d) is root\n", nq->qdisc, nq->type);
-		nss_qdisc_info("Qdisc %px dev-name %s qdisc_dev(sch)->qdisc %px, qdisc_dev(sch)->qdisc->handle %x\n", qdisc_dev(sch), qdisc_dev(sch)->name, qdisc_dev(sch)->qdisc, qdisc_dev(sch)->qdisc->handle);
-		nss_qdisc_info("Qdisc %px (sch %px) is root, sch->handle %x\n", nq->qdisc, sch, sch->handle);
 		nq->is_root = true;
 		root = sch;
 	} else {
@@ -2188,6 +2172,8 @@ int __nss_qdisc_init(struct Qdisc *sch,
 	 * This is to prevent mixing NSS and PPE qdisc with linux qdisc.
 	 */
 	if ((parent != TC_H_ROOT) && (root->ops->owner != THIS_MODULE)) {
+		nss_qdisc_warning("parent (%d) and TC_H_ROOT (%d))", parent, TC_H_ROOT);
+		nss_qdisc_warning("root->ops->owner (%px) and THIS_MODULE (%px))", root->ops->owner , THIS_MODULE);
 		nss_qdisc_warning("NSS qdisc %px (type %d) used along with non-nss qdiscs,"
 			" or the interface is currently down", nq->qdisc, nq->type);
 	}
@@ -2288,11 +2274,6 @@ int __nss_qdisc_init(struct Qdisc *sch,
 		 * If state is positive, return success
 		 */
 		if (state > 0) {
-
-			/*
-			 * Qdisc successfully initialized, add it to stats list if its not a class
-			 */
-			nss_qdisc_stats_qdisc_attach(nq);
 			return 0;
 		}

@@ -2460,15 +2441,6 @@ int __nss_qdisc_init(struct Qdisc *sch,
 		goto init_fail;
 	}
 #endif
-	/*
-	 * Initialize the stats management only for root node
-	 */
-	if (!nss_qdisc_stats_sync_many_init(nq)) {
-		nss_qdisc_error("Qdisc %px (type %d) stats sync init failed", nq->qdisc, nq->type);
-		nss_shaper_unregister_shaping(nq->nss_shaping_ctx);
-		atomic_set(&nq->state, NSS_QDISC_STATE_INIT_FAILED);
-		goto init_fail;
-	}

 	/*
 	 * Create and send the shaper assign message to the NSS interface
@@ -2498,7 +2470,7 @@ int __nss_qdisc_init(struct Qdisc *sch,
 		if (igs_put) {
 			nss_igs_module_put();
 		}
-		nss_qdisc_error("init for qdisc %x timedout!\n", nq->qos_tag);
+	nss_qdisc_error("init for qdisc %x timedout!\n", nq->qos_tag);
 		return -1;
 	}

@@ -2509,16 +2481,6 @@ int __nss_qdisc_init(struct Qdisc *sch,
 	if (state > 0) {

 		/*
-		 * Qdisc successfully initialized add it to stats list if its not a class
-		 */
-		nss_qdisc_stats_qdisc_attach(nq);
-
-		/*
-		 * Root node is successfully configured, start stats polling
-		 */
-		nss_qdisc_stats_start_polling(nq);
-
-		/*
 		 * Return if this is not a root qdisc on a bridge interface.
 		 */
 		if (!nq->is_root || !nq->is_bridge) {
@@ -2559,14 +2521,6 @@ init_fail:
 	}
 #endif

-
-	/*
-	 * Clean up stats sync objects if initialized.
-	 */
-	if (nq->stats_wq) {
-		nss_qdisc_stats_sync_many_exit(nq);
-	}
-
 	/*
 	 * Destroy any virtual interfaces created by us before returning a failure.
 	 */
@@ -2601,6 +2555,210 @@ int nss_qdisc_init(struct Qdisc *sch, st
 }

 /*
+ * nss_qdisc_basic_stats_callback()
+ *	Invoked after getting basic stats
+ */
+static void nss_qdisc_basic_stats_callback(void *app_data,
+				struct nss_if_msg *nim)
+{
+	struct nss_qdisc *nq = (struct nss_qdisc *)app_data;
+	struct Qdisc *qdisc = nq->qdisc;
+	struct gnet_stats_basic_packed *bstats;
+	struct gnet_stats_queue *qstats;
+	struct nss_shaper_node_stats_response *response;
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 13, 0))
+	atomic_t *refcnt;
+#else
+	refcount_t *refcnt;
+#endif
+
+	if (nim->cm.response != NSS_CMN_RESPONSE_ACK) {
+		nss_qdisc_warning("Qdisc %px (type %d): Receive stats FAILED - "
+			"response: type: %d\n", qdisc, nq->type,
+			nim->msg.shaper_configure.config.response_type);
+		atomic_sub(1, &nq->pending_stat_requests);
+		wake_up(&nq->wait_queue);
+		return;
+	}
+
+	response = &nim->msg.shaper_configure.config.msg.shaper_node_stats_get.response;
+
+	/*
+	 * Get the right stats pointers based on whether it is a class
+	 * or a qdisc.
+	 */
+	if (nq->is_class) {
+		bstats = &nq->bstats;
+		qstats = &nq->qstats;
+		refcnt = &nq->refcnt;
+	} else {
+		bstats = &qdisc->bstats;
+		qstats = &qdisc->qstats;
+		refcnt = &qdisc->refcnt;
+		qdisc->q.qlen = response->sn_stats.qlen_packets;
+	}
+
+	/*
+	 * Update qdisc->bstats
+	 */
+	spin_lock_bh(&nq->lock);
+	bstats->bytes += (__u64)response->sn_stats.delta.dequeued_bytes;
+	bstats->packets += response->sn_stats.delta.dequeued_packets;
+
+	/*
+	 * Update qdisc->qstats
+	 */
+	qstats->backlog = response->sn_stats.qlen_bytes;
+
+	qstats->drops += (response->sn_stats.delta.enqueued_packets_dropped +
+				response->sn_stats.delta.dequeued_packets_dropped);
+
+	/*
+	 * Update qdisc->qstats
+	 */
+	qstats->qlen = response->sn_stats.qlen_packets;
+	qstats->requeues = 0;
+	qstats->overlimits += response->sn_stats.delta.queue_overrun;
+	spin_unlock_bh(&nq->lock);
+
+	/*
+	 * Shapers that maintain additional unique statistics will process them
+	 * via a registered callback. So invoke if its been registered.
+	 */
+	if (nq->stats_cb) {
+		nq->stats_cb(nq, response);
+	}
+
+	/*
+	 * All access to nq fields below do not need lock protection. They
+	 * do not get manipulated on different thread contexts.
+	 */
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 13, 0))
+	if (atomic_read(refcnt) == 0) {
+#else
+	if (refcount_read(refcnt) == 0) {
+#endif
+		atomic_sub(1, &nq->pending_stat_requests);
+		wake_up(&nq->wait_queue);
+		return;
+	}
+
+	/*
+	 * Requests for stats again, after 1 sec.
+	 */
+	nq->stats_get_timer.expires += HZ;
+	if (nq->stats_get_timer.expires <= jiffies) {
+		nss_qdisc_info("losing time %lu, jiffies = %lu\n",
+				nq->stats_get_timer.expires, jiffies);
+		nq->stats_get_timer.expires = jiffies + HZ;
+	}
+	add_timer(&nq->stats_get_timer);
+}
+
+/*
+ * nss_qdisc_get_stats_timer_callback()
+ *	Invoked periodically to get updated stats
+ */
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
+static void nss_qdisc_get_stats_timer_callback(unsigned long int data)
+#else
+static void nss_qdisc_get_stats_timer_callback(struct timer_list *tm)
+#endif
+{
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
+	struct nss_qdisc *nq = (struct nss_qdisc *)data;
+#else
+	struct nss_qdisc *nq = from_timer(nq, tm, stats_get_timer);
+#endif
+
+	nss_tx_status_t rc;
+	struct nss_if_msg nim;
+	int msg_type;
+
+	/*
+	 * Create and send the shaper configure message to the NSS interface
+	 */
+	msg_type = nss_qdisc_get_interface_msg(nq->is_bridge, NSS_QDISC_IF_SHAPER_CONFIG);
+	nss_qdisc_msg_init(&nim, nq->nss_interface_number, msg_type, sizeof(struct nss_if_shaper_configure),
+				nss_qdisc_basic_stats_callback,
+				nq);
+	nim.msg.shaper_configure.config.request_type = NSS_SHAPER_CONFIG_TYPE_SHAPER_NODE_BASIC_STATS_GET;
+	nim.msg.shaper_configure.config.msg.shaper_node_stats_get.qos_tag = nq->qos_tag;
+	rc = nss_if_tx_msg(nq->nss_shaping_ctx, &nim);
+
+	/*
+	 * Check if we failed to send the stats request to NSS.
+	 */
+	if (rc != NSS_TX_SUCCESS) {
+		nss_qdisc_info("%px: stats fetch request dropped, causing "
+				"delay in stats fetch\n", nq->qdisc);
+
+		/*
+		 * Schedule the timer once again for re-trying. Since this is a
+		 * re-try we schedule it 100ms from now, instead of a whole second.
+		 */
+		nq->stats_get_timer.expires = jiffies + HZ/10;
+		add_timer(&nq->stats_get_timer);
+	}
+}
+
+/*
+ * nss_qdisc_start_basic_stats_polling()
+ *	Call to initiate the stats polling timer
+ */
+void nss_qdisc_start_basic_stats_polling(struct nss_qdisc *nq)
+{
+	/*
+	 * In case the stats polling timer is already
+	 * initiated, return. This can happen only when
+	 * there is a fallback from PPE to NSS qdisc.
+	 */
+	if (atomic_read(&nq->pending_stat_requests)) {
+		return;
+	}
+
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
+	init_timer(&nq->stats_get_timer);
+	nq->stats_get_timer.function = nss_qdisc_get_stats_timer_callback;
+	nq->stats_get_timer.data = (unsigned long)nq;
+#else
+	timer_setup(&nq->stats_get_timer, nss_qdisc_get_stats_timer_callback, 0);
+#endif
+
+	nq->stats_get_timer.expires = jiffies + HZ;
+	atomic_set(&nq->pending_stat_requests, 1);
+	add_timer(&nq->stats_get_timer);
+}
+
+/*
+ * nss_qdisc_stop_basic_stats_polling()
+ *	Call to stop polling of basic stats
+ */
+void nss_qdisc_stop_basic_stats_polling(struct nss_qdisc *nq)
+{
+	/*
+	 * If the timer was active, then delete timer and return.
+	 */
+	if (del_timer(&nq->stats_get_timer) > 0) {
+		/*
+		 * The timer was still active (counting down) when it was deleted.
+		 * Therefore we are sure that there are no pending stats request
+		 * for which we need to wait for. We can therefore return.
+		 */
+		return;
+	}
+
+	/*
+	 * The timer has already fired, which means we have a pending stat response.
+	 * We will have to wait until we have received the pending response.
+	 */
+	if (!wait_event_timeout(nq->wait_queue, atomic_read(&nq->pending_stat_requests) == 0,
+				NSS_QDISC_COMMAND_TIMEOUT)) {
+		nss_qdisc_error("Stats request command for %x timedout!\n", nq->qos_tag);
+	}
+}
+
+/*
  * nss_qdisc_gnet_stats_copy_basic()
  *  Wrapper around gnet_stats_copy_basic()
  */
@@ -2876,10 +3034,6 @@ static int __init nss_qdisc_module_init(
 		return ret;
 	nss_qdisc_info("nss qdisc device notifiers registered\n");

-	if (!nss_qdisc_stats_work_queue_init()) {
-		nss_qdisc_error("Failed to initialized stats workqueue thread\n");
-	}
-
 #if defined(NSS_QDISC_PPE_SUPPORT)
 	nss_ppe_port_res_alloc();
 	nss_qdisc_info("nss ppe qdsic configured");
@@ -2899,8 +3053,6 @@ static void __exit nss_qdisc_module_exit
 	}
 #endif

-	nss_qdisc_stats_work_queue_exit();
-
 	unregister_qdisc(&nss_pfifo_qdisc_ops);
 	nss_qdisc_info("nsspfifo unregistered\n");

--- a/nss_qdisc/nss_qdisc.h
+++ b/nss_qdisc/nss_qdisc.h
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2018, 2020-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014-2018, 2020 The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -35,17 +35,12 @@
 #if defined(NSS_QDISC_PPE_SUPPORT)
 #include "nss_ppe.h"
 #endif
-#include "nss_qdisc_stats.h"
-#include "nss_qdisc_htable.h"

 #define NSS_QDISC_DEBUG_LEVEL_ERROR 1
 #define NSS_QDISC_DEBUG_LEVEL_WARN 2
 #define NSS_QDISC_DEBUG_LEVEL_INFO 3
 #define NSS_QDISC_DEBUG_LEVEL_TRACE 4

-#define NSS_QDISC_COMMAND_TIMEOUT (10*HZ) /* We set 10sec to be the command */
-					   /* timeout value for messages */
-
 /*
  * Debug message for module init and exit
  */
@@ -92,9 +87,6 @@
 #endif
 #endif

-#define NSS_QDISC_STATS_SYNC_MANY_PERIOD msecs_to_jiffies(1000)
-#define NSS_QDISC_STATS_SYNC_MANY_UDELAY 5000     /* Delay for 5 ms */
-
 /*
  * State values
  */
@@ -144,27 +136,6 @@ struct nss_qdisc;
 typedef void (*nss_qdisc_stats_callback_t)(struct nss_qdisc *nq, struct nss_shaper_node_stats_response *response);
 typedef void (*nss_qdisc_configure_callback_t)(struct nss_qdisc *nq, struct nss_shaper_configure *response);

-
-/*
- * Qdisc stats sync info object
- */
-struct nss_qdisc_stats_wq {
-	struct nss_qdisc *nq;			/* Pointer to root nss_qdisc */
-	struct list_head stats_list;		/* List head stats sync management work */
-	struct nss_if_msg stats_sync_req_msg;	/* FW sync message */
-	struct timer_list stats_get_timer;	/* Timer used to start fresh iter */
-	unsigned long int next_req_time;	/* Time at which next sync iteration starts */
-	unsigned long int stats_request_success;	/* Number of success stats request */
-	unsigned long int stats_request_fail;		/* Number of failed stats request */
-	unsigned long int stats_request_nack;		/* Number of NACK'd stats request */
-	atomic_t pending_stat_work;		/* Pending work queue status */
-	atomic_t pending_stat_resp;		/* Pending statistics response from FW */
-	bool stats_polling_stopped;		/* True when polling is stopped due to qdisc delete */
-	wait_queue_head_t stats_work_waitqueue;	/* Wait queue to wait on work queue processing */
-	wait_queue_head_t stats_resp_waitqueue;	/* Wait queue used to wait on response from the NSS */
-	struct nss_qdisc_htable nqt;		/* Struct to manage hash table of Qdiscs for stats */
-};
-
 struct nss_qdisc {
 	struct Qdisc *qdisc;			/* Handy pointer back to containing qdisc */
 	struct nss_qdisc *parent;		/* Pointer to parent nss qdisc */
@@ -226,8 +197,6 @@ struct nss_qdisc {
 #endif
 	struct timer_list stats_get_timer;	/* Timer used to poll for stats */
 	atomic_t pending_stat_requests;		/* Number of pending stats responses */
-	struct nss_qdisc_stats_wq *stats_wq;	/* Stats info and state work object */
-	struct hlist_node hlist;		/* Hlist node for managing stats hash list */
 	wait_queue_head_t wait_queue;		/* Wait queue used to wait on responses from the NSS */
 	spinlock_t lock;			/* Lock to protect the nss qdisc structure */
 	uint16_t mode;				/* Mode of Qdisc/class */
@@ -460,6 +429,18 @@ extern int nss_qdisc_init(struct Qdisc *
 		void *extack);

 /*
+ * nss_qdisc_start_basic_stats_polling()
+ *	Call to initiate the stats polling timer
+ */
+extern void nss_qdisc_start_basic_stats_polling(struct nss_qdisc *nq);
+
+/*
+ * nss_qdisc_stop_basic_stats_polling()
+ *	Call to stop polling of basic stats
+ */
+extern void nss_qdisc_stop_basic_stats_polling(struct nss_qdisc *nq);
+
+/*
  * nss_qdisc_gnet_stats_copy_basic()
  *  Wrapper around gnet_stats_copy_basic()
  */
@@ -505,15 +486,3 @@ extern unsigned long nss_qdisc_tcf_bind(
  *	Unbind the filter from the qdisc.
  */
 extern void nss_qdisc_tcf_unbind(struct Qdisc *sch, unsigned long arg);
-
-/*
- * nss_qdisc_get_interface_msg()
- *	Returns the correct message that needs to be sent down to the NSS interface.
- */
-extern int nss_qdisc_get_interface_msg(bool is_bridge, uint32_t msg_type);
-
-/*
- * nss_qdisc_msg_init()
- *      Initialize the qdisc specific message
- */
-extern void nss_qdisc_msg_init(struct nss_if_msg *nim, uint16_t if_num, uint32_t msg_type, uint32_t len, nss_if_msg_callback_t cb, void *app_data);
--- a/nss_qdisc/nss_tbl.c
+++ b/nss_qdisc/nss_tbl.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2017, 2019-2021, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014-2017, 2019-2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -88,7 +88,12 @@ static void nss_tbl_destroy(struct Qdisc
 	/*
 	 * Now we can destroy our child qdisc
 	 */
-	nss_qdisc_put(q->qdisc);
+	 nss_qdisc_put(q->qdisc);
+
+	/*
+	 * Stop the polling of basic stats and destroy qdisc.
+	 */
+	nss_qdisc_stop_basic_stats_polling(&q->nq);
 	nss_qdisc_destroy(&q->nq);
 }

@@ -287,6 +292,11 @@ static int nss_tbl_init(struct Qdisc *sc
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(&q->nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_wred.c
+++ b/nss_qdisc/nss_wred.c
@@ -1,6 +1,6 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2017, 2020-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2014-2017, 2020 The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
@@ -108,6 +108,11 @@ static void nss_wred_destroy(struct Qdis
 {
 	struct nss_qdisc *nq = (struct nss_qdisc *)qdisc_priv(sch);

+	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(nq);
+
 	nss_qdisc_destroy(nq);
 	nss_qdisc_info("nsswred destroyed");
 }
@@ -370,6 +375,11 @@ static int nss_wred_init(struct Qdisc *s
 		return -EINVAL;
 	}

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(nq);
+
 	return 0;
 }

--- a/nss_qdisc/nss_wrr.c
+++ b/nss_qdisc/nss_wrr.c
@@ -1,13 +1,9 @@
 /*
  **************************************************************************
- * Copyright (c) 2014-2017, 2019-2021, The Linux Foundation. All rights reserved.
- *
- * Copyright (c) 2022, Qualcomm Innovation Center, Inc. All rights reserved.
- *
+ * Copyright (c) 2014-2017, 2019-2020, The Linux Foundation. All rights reserved.
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
  * above copyright notice and this permission notice appear in all copies.
- *
  * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
@@ -91,6 +87,11 @@ static void nss_wrr_destroy_class(struct
 	 nss_qdisc_put(cl->qdisc);

 	/*
+	 * Stop the stats polling timer and free class
+	 */
+	nss_qdisc_stop_basic_stats_polling(&cl->nq);
+
+	/*
 	 * Destroy the shaper in NSS
 	 */
 	nss_qdisc_destroy(&cl->nq);
@@ -351,6 +352,11 @@ static int nss_wrr_change_class(struct Q
 		 */
 		qdisc_class_hash_grow(sch, &q->clhash);

+		/*
+		 * Start the stats polling timer
+		 */
+		nss_qdisc_start_basic_stats_polling(&cl->nq);
+
 		nss_qdisc_info("Class %u successfully allocated\n", classid);
 	}

@@ -400,7 +406,11 @@ failure:
 	return -EINVAL;
 }

+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0))
+static int nss_wrr_delete_class(struct Qdisc *sch, unsigned long arg, struct netlink_ext_ack *extack)
+#else
 static int nss_wrr_delete_class(struct Qdisc *sch, unsigned long arg)
+#endif
 {
 	struct nss_wrr_sched_data *q = qdisc_priv(sch);
 	struct nss_wrr_class_data *cl = (struct nss_wrr_class_data *)arg;
@@ -408,14 +418,11 @@ static int nss_wrr_delete_class(struct Q
 	int refcnt;

 	/*
-	 * If the class is a root class or has a child qdisc attached
-	 * we do not support deleting it.
+	 * Since all classes are leaf nodes in our case, we dont have to make
+	 * that check.
 	 */
-	if ((cl == &q->root) || (cl->qdisc != &noop_qdisc)) {
-		nss_qdisc_warning("Cannot delete wrr class %x as it is the "
-				  "root class or has a child qdisc attached\n", cl->nq.qos_tag);
+	if (cl == &q->root)
 		return -EBUSY;
-	}

 	/*
 	 * The message to NSS should be sent to the parent of this class
@@ -429,24 +436,16 @@ static int nss_wrr_delete_class(struct Q
 	}

 	sch_tree_lock(sch);
+	qdisc_reset(cl->qdisc);
 	qdisc_class_hash_remove(&q->clhash, &cl->cl_common);

 	refcnt = nss_qdisc_atomic_sub_return(&cl->nq);

 	sch_tree_unlock(sch);
-
-	/*
-	 * For 5.4 and above kernels, calling nss_htb_destroy_class
-	 * explicitly as there is no put_class which would have called
-	 * nss_wrr_destroy_class when refcnt becomes zero.
-	 */
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0))
-	nss_wrr_destroy_class(sch, cl);
-#else
 	if (!refcnt) {
 		nss_qdisc_error("Reference count should not be zero for class %px\n", cl);
 	}
-#endif
+
 	return 0;
 }

@@ -722,6 +721,11 @@ static int nss_wrr_init_qdisc(struct Qdi

 	nss_qdisc_info("Nsswrr initialized - handle %x parent %x\n", sch->handle, sch->parent);

+	/*
+	 * Start the stats polling timer
+	 */
+	nss_qdisc_start_basic_stats_polling(&q->nq);
+
 	return 0;
 }

@@ -760,11 +764,6 @@ static int nss_wrr_change_qdisc(struct Q

 static void nss_wrr_reset_class(struct nss_wrr_class_data *cl)
 {
-	if (cl->qdisc == &noop_qdisc) {
-		nss_qdisc_trace("Class %x has no child qdisc to reset\n", cl->nq.qos_tag);
-		return;
-	}
-
 	nss_qdisc_reset(cl->qdisc);
 	nss_qdisc_info("Nsswrr class resetted %px\n", cl->qdisc);
 }
@@ -837,6 +836,11 @@ static void nss_wrr_destroy_qdisc(struct
 	qdisc_class_hash_destroy(&q->clhash);

 	/*
+	 * Stop the polling of basic stats
+	 */
+	nss_qdisc_stop_basic_stats_polling(&q->nq);
+
+	/*
 	 * Now we can go ahead and destroy the qdisc.
 	 * Note: We dont have to detach ourself from our parent because this
 	 *	 will be taken care of by the graft call.
