--- a/tunipip6/Makefile
+++ b/tunipip6/Makefile
@@ -2,7 +2,6 @@
 ccflags-y += -I$(obj)/../exports -I$(obj)/..
 ccflags-y += -DNSS_CLIENT_BUILD_ID="$(BUILD_ID)"
 ccflags-y += -DNSS_TUNIPIP6_DEBUG_LEVEL=0
-ccflags-y += -Wall -Werror
 obj-m += qca-nss-tunipip6.o
 qca-nss-tunipip6-objs := nss_connmgr_tunipip6.o nss_connmgr_tunipip6_sysctl.o nss_connmgr_tunipip6_stats.o
 ifneq ($(findstring 4.4, $(KERNELVERSION)),)
--- a/tunipip6/nss_connmgr_tunipip6.c
+++ b/tunipip6/nss_connmgr_tunipip6.c
@@ -194,7 +194,6 @@ static void nss_tunipip6_encap_exception
 		skb->pkt_type = PACKET_HOST;
 		skb->skb_iif = dev->ifindex;
 		skb->ip_summed = CHECKSUM_NONE;
-		ip_rt_put(rt);
 		netif_receive_skb(skb);
 		return;
 	}
@@ -348,7 +347,7 @@ static void nss_tunipip6_update_dev_stat
 {
 	struct pcpu_sw_netstats stats;
 	enum nss_dynamic_interface_type interface_type;
-	struct nss_tunipip6_stats_sync_msg *sync_stats = (struct nss_tunipip6_stats_sync_msg *)&tnlmsg->msg.stats;
+	struct nss_tunipip6_stats_sync_msg *sync_stats = (struct nss_tunipip6_stats_sync_msg *)&tnlmsg->msg.stats_sync;

 	interface_type = nss_dynamic_interface_get_type(nss_tunipip6_get_context(), tnlmsg->cm.interface);


@@ -381,7 +380,7 @@ void nss_tunipip6_event_receive(void *if
 	netdev = (struct net_device *)if_ctx;

 	switch (tnlmsg->cm.type) {
-	case NSS_TUNIPIP6_STATS_SYNC:
+	case NSS_TUNIPIP6_RX_STATS_SYNC:
 		/*
 		 * Update netdevice statistics.
 		 */

@@ -627,7 +626,6 @@ configure_tunnel:
 	tnlcreate->ttl_inherit = tnlcfg->ttl_inherit;
 	tnlcreate->tos_inherit = tnlcfg->tos_inherit;
 	tnlcreate->frag_id_update = tnlcfg->frag_id_update;
-	tnlcreate->fmr_max = tnlcfg->fmr_max;

 	/*
 	 * Set "draft03" based on "tunnel_type". draft03 should be


--- a/tunipip6/nss_connmgr_tunipip6_stats.c
+++ b/tunipip6/nss_connmgr_tunipip6_stats.c
@@ -132,7 +132,7 @@ static void nss_tunipip6_stats_update(ui
 void nss_tunipip6_stats_sync(struct net_device *dev, struct nss_tunipip6_msg *ntm)
 {
 	uint32_t ifnum = ntm->cm.interface;
-	struct nss_tunipip6_stats_sync_msg *stats = &ntm->msg.stats;
+	struct nss_tunipip6_stats_sync_msg *stats = &ntm->msg.stats_sync;
 	struct nss_tunipip6_instance *ntii;
 	struct nss_tunipip6_stats *s;

@@ -147,7 +147,6 @@ void nss_tunipip6_stats_sync(struct net_
 	s = &ntii->stats;
 	if (ntii->inner_ifnum == ifnum) {
 		nss_tunipip6_stats_update(s->inner_stats, stats);
-		s->inner_stats[NSS_TUNIPIP6_STATS_CONFIG_ENCAP_TOTAL_FMR] = stats->tun_stats.encap.cfg.total_fmr;
 	} else if (ntii->outer_ifnum == ifnum) {
 		nss_tunipip6_stats_update(s->outer_stats, stats);
 	} else {
--- a/tunipip6/nss_connmgr_tunipip6_sysctl.c
+++ b/tunipip6/nss_connmgr_tunipip6_sysctl.c
@@ -449,6 +449,7 @@ static struct ctl_table nss_tunipip6_tab
 	{ }
 };

+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
 static struct ctl_table nss_tunipip6_root_dir[] = {
 	{
 		.procname		= "ipip6",
@@ -475,6 +476,7 @@ static struct ctl_table nss_tunipip6_roo
 	},
 	{ }
 };
+#endif

 static struct ctl_table_header *nss_tunipip6_ctl_header;

@@ -483,7 +485,11 @@ static struct ctl_table_header *nss_tuni
  * 	Register command line interface for tunipip6.
  */
 bool nss_tunipip6_sysctl_register(void) {
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(6, 6, 0))
+	nss_tunipip6_ctl_header = register_sysctl("drv/nss/ipip6", nss_tunipip6_table);
+#else
 	nss_tunipip6_ctl_header = register_sysctl_table(nss_tunipip6_root);
+#endif
 	if (!nss_tunipip6_ctl_header) {
 		return false;
 	}
--- a/vlan/nss_vlan_mgr.c
+++ b/vlan/nss_vlan_mgr.c
@@ -1557,6 +1557,7 @@ static struct ctl_table nss_vlan_table[]
 	{ }
 };

+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
 /*
  * nss_vlan sysctl dir
  */
@@ -1580,6 +1581,7 @@ static struct ctl_table nss_vlan_root_di
 	},
 	{ }
 };
+#endif

 /*
  * nss_vlan_mgr_add_bond_slave()
@@ -1920,7 +1922,11 @@ int __init nss_vlan_mgr_init_module(void
 	vlan_mgr_ctx.stpid = ETH_P_8021Q;

 #ifdef NSS_VLAN_MGR_PPE_SUPPORT
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(6, 6, 0))
+	vlan_mgr_ctx.sys_hdr = register_sysctl("nss/vlan_client", nss_vlan_table);
+#else
 	vlan_mgr_ctx.sys_hdr = register_sysctl_table(nss_vlan_root_dir);
+#endif
 	if (!vlan_mgr_ctx.sys_hdr) {
 		nss_vlan_mgr_warn("Unabled to register sysctl table for vlan manager\n");
 		return -EFAULT;
