--- a/pvxlanmgr/nss_pvxlanmgr.c
+++ b/pvxlanmgr/nss_pvxlanmgr.c
@@ -177,7 +177,11 @@ static struct rtnl_link_stats64 *nss_pvx
 	 * Netdev seems to be incrementing rx_dropped because we don't give IP header.
 	 * So reset it as it's of no use for us.
 	 */
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(6, 1, 0))
+	atomic_long_set(&(dev)->stats.__rx_dropped, 0);
+#else
 	atomic_long_set(&dev->rx_dropped, 0);
+#endif
 	priv = netdev_priv(dev);
 	memset(stats, 0, sizeof(struct rtnl_link_stats64));
 	memcpy(stats, &priv->stats, sizeof(struct rtnl_link_stats64));
@@ -305,7 +309,11 @@ static void nss_pvxlanmgr_dummy_netdev_s
 	dev->priv_destructor = NULL;
 #endif

+#if (LINUX_VERSION_CODE > KERNEL_VERSION(6, 1, 0))
+	memcpy((void *)dev->dev_addr, "\x00\x00\x00\x00\x00\x00", dev->addr_len);
+#else
 	memcpy(dev->dev_addr, "\x00\x00\x00\x00\x00\x00", dev->addr_len);
+#endif
 	memset(dev->broadcast, 0xff, dev->addr_len);
 	memcpy(dev->perm_addr, dev->dev_addr, dev->addr_len);
 }
